src/ray/core_worker/task_manager.cc

```diff
@@ -169,26 +186,38 @@ void TaskManager::ShutdownIfNeeded() {
   }
 }
 
-void TaskManager::RemoveSubmittedTaskReferences(const std::vector<ObjectID> &object_ids) {
+void TaskManager::RemoveSubmittedTaskReferences(
+    const std::vector<ObjectID> &object_ids, const rpc::Address &worker_addr,
+    const ReferenceCounter::ReferenceTableProto &borrowed_refs) {
   std::vector<ObjectID> deleted;
-  reference_counter_->RemoveSubmittedTaskReferences(object_ids, &deleted);
+  reference_counter_->RemoveSubmittedTaskReferences(object_ids, worker_addr,
+                                                    borrowed_refs, &deleted);
   in_memory_store_->Delete(deleted);
 }
 
-void TaskManager::OnTaskDependenciesInlined(const std::vector<ObjectID> &object_ids) {
-  RemoveSubmittedTaskReferences(object_ids);
+void TaskManager::OnTaskDependenciesInlined(
+    const std::vector<ObjectID> &inlined_dependency_ids,
+    const std::vector<ObjectID> &contained_ids) {
+  reference_counter_->AddSubmittedTaskReferences(contained_ids);
+  RemoveSubmittedTaskReferences(inlined_dependency_ids);
 }
 
-void TaskManager::RemovePlasmaSubmittedTaskReferences(TaskSpecification &spec) {
+void TaskManager::RemoveFinishedTaskReferences(
+    TaskSpecification &spec, const rpc::Address &borrower_addr,
+    const ReferenceCounter::ReferenceTableProto &borrowed_refs) {
   std::vector<ObjectID> plasma_dependencies;
   for (size_t i = 0; i < spec.NumArgs(); i++) {
-    auto count = spec.ArgIdCount(i);
-    if (count > 0) {
-      const auto &id = spec.ArgId(i, 0);
-      plasma_dependencies.push_back(id);
+    if (spec.ArgByRef(i)) {
+      for (size_t j = 0; j < spec.ArgIdCount(i); j++) {
+        plasma_dependencies.push_back(spec.ArgId(i, j));
+      }
+    } else {
+      const auto &inlined_ids = spec.ArgInlinedIds(i);
+      plasma_dependencies.insert(plasma_dependencies.end(), inlined_ids.begin(),
+                                 inlined_ids.end());
     }
   }
-  RemoveSubmittedTaskReferences(plasma_dependencies);
+  RemoveSubmittedTaskReferences(plasma_dependencies, borrower_addr, borrowed_refs);
```

Comment:
Can we use the same naming for this and `RemoveFinishedTaskReferences`? It's a bit confusing right now that one refers to them as `Submitted` and one as `Finished`. Either one is fine but slight preference for `Finished`

Reply:
I changed the naming to make it clearer that one is called when the task has been submitted but not finished, and the other should be called only when the task has finished. I prefer to keep them separate.